README.md

z3c.rest_publisher Package
==========================

Overview
--------

This product registers Zope-traverses (IBrowserPublisher) to implement basic REST requests in a simpler way.
This is done by registering traversable objects (IBrowserPublisher) to represent each method or level of the REST API.
No documents or folders will be accessed directly. It is necessary to create classes for each information to be accessed by the API.

For example, in the following request:

    curl http://localhost:8080/api/members/my_user

where
 - api - It's a view (inherited from APIBase) registered for ROOT
 - members - It's a object inherited from APIBase included in "api"
 - myuser - It's a object inherited from APIBase included in "members"

Install
-------

pip install z3c.rest_publisher

Configuration
-------------

1 - API Root
The starting point for the API is a traditional view registered in Zope ZCML. e.g.

```xml
<view
name="api"
for="zope.location.interfaces.IRoot"
type="zope.publisher.interfaces.browser.IBrowserRequest"
provides="zope.publisher.interfaces.browser.IBrowserPublisher"
factory=".rest_api.APIRoot"
permission="zope.Public"
allowed_attributes="publishTraverse browserDefault __call__">
</view>
```

Authentication
--------------

1 - Zope-Authentication
You can use traditional Zope systems for authentication, for example by changing permissions for the view

    permission="zope.Public"
    or
    permission="zope.ManageContent"

2 - Custom authentication

If you wish, you can include customized authentication for the API. Just overwrite the "check_authentication" method in the class.
For example, to implement a basic authentication:

    class APIRoot(APIBase):
        def check_authentication(self):
            user, pwd = self.request._authUserPW()
            return user == 'demo' and pwd == 'demo'

To test:

    curl --user demo:demo  http://localhost:8080/api/

REST functions
--------------

There are two ways to add a REST function to the Object.

1 - REST function as a method
Include a method in the format {HTTP-Method}_{Name} (lower-case), e.g.

    class APIRoot(APIBase):
        def get_list_admins(self):
            return [{'id': '1', 'firstname': 'Alberto', 'lastname': 'Santos-Dumont'},
                    {'id': '2', 'firstname': 'Edson', 'lastname': 'Arantes do Nascimento'}]
        def put_list_admins(self):
            return self.get_list_admins()

To test:

    curl -X GET http://localhost:8080/api/list_admins
    curl -X PUT http://localhost:8080/api/list_admins

2 - Concatenated REST objects
It is possible to create a new REST object and concatenate it with the current object.
This is the equivalent implementation from the previous method.

    class APIListAdmins(APIBase):
        def get(self):
            return [{'id': 'user1', 'firstname': 'Alberto', 'lastname': 'Santos-Dumont'},
                    {'id': 'user2', 'firstname': 'Edson', 'lastname': 'Arantes do Nascimento'}]
        def put(self):
            return self.get()

    class APIRoot(APIBase):
        content = {'list_admins': APIListAdmins}

To test:

    curl -X GET http://localhost:8080/api/list_admins
    curl -X PUT http://localhost:8080/api/list_admins

2.1 - REST function for objects (catch all)
To implement a "generic" traverse to access specific database objects, use the wildcard "*".

    class APIUSer(APIBase):
        def get(self):
            if self.name == 'user1'
                return {'id': 'user1', 'firstname': 'Alberto', 'lastname': 'Santos-Dumont'}
            elif self.name == 'user2':
                return {'id': 'user2', 'firstname': 'Edson', 'lastname': 'Arantes do Nascimento'}
            else:
                self.request.response.setStatus(404)
                return "User not found"

    class APIMembers(APIBase):
        content = {'*': APIUSer}

    class APIRoot(APIBase):
        content = {'members': APIListAdmins}


Be aware that curl just send data with POST methods. This is a limitation discussed in some posts.
To test these methods with curl, use the 'querystring_verb_name' option as documented. e.g.
querystring_verb_name=verb
curl -X POST --user demo:demo  http://127.0.0.1:9095/testapi/companies/company1/sectors/sector2/users/user4?verb=PATCH  --data-raw '{"firstname": "New Name"}'

To test:

    curl -X GET http://localhost:8080/api/members/user1
    
(Check more examples in example/README.txt)  

API documentation
-----------------

The class RestDoc provide a simple built-in documentation of all REST-APi methods. 
The default url is '[API_ROOT]/help' but you can change the property 'doc_endpoint' (APIBase) if you desire.


Example
-------

Please see "z3c.rest_publisher/example" for an example.


CHANGES.txt
Changelog
=========

0.9.1 (15.07.20)
----------------

- Redesign code.
- Improved the documentation
- Added a complete example

0.8 (08.05.20)
--------------

- Added the built-in documentation (default: /help) for the REST API
- Fixed interfaces


0.4 (01.05.20)
--------------

- Initial release



z3c/rest_publisher/tests.py
import unittest

from zope.testing import doctestunit
from zope.component import testing
from Testing import ZopeTestCase as ztc

def test_suite():
    return unittest.TestSuite([

        # Unit tests for your API
        doctestunit.DocFileSuite(
            'README.md', package='z3c.rest_publisher',
            setUp=testing.setUp, tearDown=testing.tearDown),

        #TODO

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


z3c/rest_publisher/base.py
import datetime
import json
import logging

from wcore import interfaces
from zope.browserpage import ViewPageTemplateFile
from zope.component import adapts, queryMultiAdapter
from zope.interface import implements
from zope.publisher.browser import BrowserPage
from zope.publisher.browser import BrowserPage
from zope.publisher.interfaces import NotFound
from zope.publisher.interfaces import TraversalException
from zope.publisher.interfaces.browser import IBrowserRequest, IBrowserPublisher

logger = logging.getLogger(__name__)


class APIBase(object):
    """ Base class for REST Objects """
    adapts(IBrowserRequest)
    implements(IBrowserPublisher)

    # False (default) is used to avoid returning the default error page (HTML).
    raise_exceptions = False
    # it is possible to inform the verb in the query-string in case you can't use http-methods.
    # e.g. If querystring_verb_name='verb', send a POST request and add something like ?verb=PATCH in the query-string.
    querystring_verb_name = ""
    # Make this API public - It will not check the custom authentication
    public = False
    # All methods supported by this API
    allowed_methods = ['get', 'post', 'put', 'delete', 'patch']

    def __init__(self, context, request, name, parent_api_obj=None):
        self.context = context
        self.request = request
        self.name = name
        self.parent_api_obj = parent_api_obj

    def check_authentication(self, name):
        """  Overwrite this method to add a custom authentication for your REST API """
        return True

    @staticmethod
    def format_unsupported_value(value):
        """ Use this static function if there is a value that is not supported by the standard 'format_output'
        function """
        if isinstance(value, datetime.datetime):
            return value.isoformat()
        elif isinstance(value, datetime.date):
            return value.isoformat()
        elif value is None:
            return ''
        else:
            return str(value)

    def format_output(self, data):
        """ format the output. default: json  """
        return json.dumps(data, default=self.format_unsupported_value)

    def browserDefault(self, request):
        return self, ()

    def error_verb(self, ex):
        """ Overwrite this function if you want specific action on verb errors """
        logger.error(ex, exc_info=True)
        if not self.raise_exceptions:
            self.request.response.setStatus(500)
            return "Server error (verb) - %s" % ex
        else:
            raise

    def error_traverse(self, ex, request, name):
        """ Overwrite this function if you want specific action on traverse errors """
        logger.error(ex, exc_info=True)
        raise

    def __call__(self):
        try:
            method = self.request['REQUEST_METHOD'].lower()
            if self.querystring_verb_name:
                method = self.request.get(self.querystring_verb_name, method).lower()

            if method not in self.allowed_methods:
                self.request.response.setStatus(403)
                return "The HTTP-method '%s' is not supported" % method

            if not self.public and not self.check_authentication(method):
                self.request.response.setStatus(403)
                return "Access denied"

            # Search for a method with the same name as REQUEST_METHOD and call it
            fnc = "verb_%s" % method
            if hasattr(self, fnc):
                res = getattr(self, fnc)()
                return self.format_output(res)
            else:
                self.request.response.setStatus(500)
                return "The HTTP-method '%s' is not supported here" % method
        except Exception as ex:
            return self.error_verb(ex)

    def publishTraverse(self, request, name):
        try:
            if not self.public and not self.check_authentication(name):
                raise TraversalException("Access denied")

            # Traverse 1 - Call traverse_{Name} to get one API-Object with the name {Name}
            method = self.request['REQUEST_METHOD'].lower()
            func = "traverse_{name}".format(name=name)
            if hasattr(self, func):
                obj = getattr(self, func)(request, name)
                return obj

            # Traverse 2 - Call traverse_NAME to return a child-object with the key = {name}
            func = "traverse_NAME"
            if hasattr(self, func):
                obj = getattr(self, func)(request, name)
                return obj

            # Traverse 3 - Fall back to views
            view = queryMultiAdapter((self.context, request), name=name)
            if view is not None:
                return view

            # Traverse 4 - If nothing found, this method will be called
            func = "traverse_DEFAULT"
            if hasattr(self, func):
                obj = getattr(self, func)(request, name)
                return obj
            # Return a 404 Not Found error page
            raise NotFound(self.context, name, request)
        except NotFound:
            raise
        except Exception as ex:
            return self.error_traverse(ex, request, name)


z3c/rest_publisher/example/rest_api_example.py
import json
import os
from z3c.rest_publisher.base import APIBase
from zope.publisher.interfaces import NotFound

# Read the DATA from file
f = open(os.path.join(os.path.dirname(__file__), "DB.json"))
DB = json.load(f)[0]


# Example with a object
class Headquarter(object):
    name = "The companie Group"
    city = "Neverland"
    state = "SC"
    mail = "info@thecompnia.com"


headquarter = Headquarter()


class SecureAPIBase(APIBase):
    """ A API Base class with custom authentication """
    querystring_verb_name = "verb"

    def check_authentication(self, name):
        """ Authenticate requests with basic-authentication user:demo, pwd:demo """
        return True
        # TODO
        cred = self.request._authUserPW()
        if cred and len(cred) == 2:
            user, pwd = self.request._authUserPW()
            if user == 'demo' and pwd == 'demo':
                return True


class APIUser(SecureAPIBase):
    """ /api/companies/IdX/sectors/IdY/users/IdZ - API for user IdZ in sector IdY in the company IdX """

    def verb_get(self):
        """ List details about this user """
        user = self.context
        return dict([(k, v) for (k, v) in user.items() if k in ["id", "firstname", "lastname", "mail", "admin"]])

    def verb_delete(self):
        """ Delete the user """
        user = self.context
        users_api = self.parent_api_obj
        users = users_api.context
        users.remove(user)
        return "user removed"

    def verb_patch(self):
        """ Update some attributes from user """
        user = self.context
        import pdb;
        pdb.set_trace()
        for k, v in self.request.form.items():
            user[k] = v
        return self.verb_get()

    def verb_post(self):
        """ Update all attributes from user """
        return self.verb_patch()


class APIUsers(SecureAPIBase):
    """ /api/companies/IdX/sectors/IdY/users - API for all users in sector IdY in the company IdX """

    def traverse_NAME(self, request, name):
        """ search for a user with id = name and return the API for the user object  """
        users = self.context
        user = [s for s in users if s['id'] == name]
        if user:
            return APIUser(context=user[0], request=request, name=name, parent_api_obj=self)
        else:
            raise NotFound(self.context, name, request)

    def verb_get(self):
        """ Get data from all users """
        users = self.context
        ret = []
        for u in users:
            ret.append({"id": u["id"], "firstname": u["firstname"], "admin": u["admin"]})
        return ret

    def verb_post(self):
        """ Add a new user """
        data = self.request.form
        users = self.context
        data["admin"] = data.get("admin", "FALSE").upper() == "TRUE"
        users.append(data)
        return self.verb_get()


class APISector(SecureAPIBase):
    """ /api/companies/IdX/sectors/IdY - API for sector IdY in the company IdX
    """

    def traverse_users(self, request, name):
        """ return the API for members objects  """
        sector = self.context
        return APIUsers(context=sector["users"], request=request, name=name, parent_api_obj=self)

    def verb_get(self):
        """ List details about this sector """
        sector = self.context
        return dict([(k, v) for (k, v) in sector.items() if k in ["id", "title"]])


class APISectors(SecureAPIBase):
    """ /api/companies/IdX/sectors/IdY - API for all sectors in the company IdX
    """

    def traverse_NAME(self, request, name):
        """ search for a sector with id = name and return the API for the sector object  """
        sectors = self.context
        sector = [s for s in sectors if s['id'] == name]
        if sector:
            return APISector(context=sector[0], request=request, name=name, parent_api_obj=self)
        else:
            raise NotFound(self.context, name, request)

    def verb_get(self):
        """ Get data from all sectors """
        sectors = self.context
        ret = []
        for c in sectors:
            ret.append({"id": c["id"], "title": c["title"], "users": [s["id"] for s in c["users"]]})
        return ret


class APICompany(SecureAPIBase):
    """ /api/companies/IdX - API for the company IdX
    """

    def traverse_sectors(self, request, name):
        """ return the API for sectors  """
        company = self.context
        return APISectors(context=company["sectors"], request=request, name=name, parent_api_obj=self)

    def verb_get(self):
        """ List details about this company """
        company = self.context
        return dict([(k, v) for (k, v) in company.items() if k in ["id", "title", "telephone", "mail"]])


class APICompanies(SecureAPIBase):
    """ /api/companies/ - API for all companies
    """

    def traverse_NAME(self, request, name):
        """ search for a company with id = name and return the API for the company object  """
        dbroot = self.context
        company = [c for c in dbroot['companies'] if c['id'] == name]
        if company:
            return APICompany(context=company[0], request=request, name=name, parent_api_obj=self)
        else:
            raise NotFound(self.context, name, request)

    def verb_get(self):
        """ Get data from all companies """
        dbroot = self.context
        ret = []
        for c in dbroot['companies']:
            ret.append({"id": c["id"], "title": c["title"], "sectors": [s["id"] for s in c["sectors"]]})
        return ret


class APIHeadquarter(SecureAPIBase):
    """ /api/headquarter - API for headquarter object
    """

    def verb_get(self):
        """ Get data from object headquarter """
        headquarter_obj = self.context
        ret = {}
        for obj_attr in ['name', 'city', 'state', 'mail']:
            ret[obj_attr] = getattr(headquarter_obj, obj_attr, "[Not_Found]")
        return ret


class APIRoot(SecureAPIBase):
    """ REST - Access the API as a view
    Check the README.txt for some examples how to test with curl, e.g.
    curl -X GET --user demo:demo  http://localhost:8080/api/companies
    """
    public = True

    def traverse_headquarter(self, request, name):
        """
        Return information from headquarter
        :return: APIHeadquarter
        """
        return APIHeadquarter(context=headquarter, request=request, name=name, parent_api_obj=self)

    def traverse_companies(self, request, name):
        return APICompanies(context=DB, request=request, name=name, parent_api_obj=self)

    def __init__(self, context, request):
        self.name = "testapi"
        self.context = context
        self.request = request


z3c/rest_publisher/example/rest_api.py

from z3c.rest_publisher.base import APIBase

USERS_EXAMPLE = [
    {'id': 'id1', 'firstname': 'Alberto', 'lastname': 'Santos-Dumont', 'mail': 'alberto@airplane.com', 'admin': True},
    {'id': 'id2', 'firstname': 'Edson', 'lastname': 'Arantes do Nascimento', 'mail': 'pele@futebol.com', 'admin': False},
    {'id': 'id3', 'firstname': 'Alberto', 'lastname': 'Santos Dummont', 'mail': 'alberto@airplane.com', 'admin': True},
]

class APIMember(APIBase):
    """ /api/members/{user_id} - Access information about a specific member """

    def get(self):
        """ return the data from user """
        for user in USERS_EXAMPLE:
            if self.name == user[id]:
                return user
        self.request.response.setStatus(404)
        return {'return_code': 1, 'errors': ((1, 'Member not found!'),)}


class APIMembers(APIBase):
    """ /api/members - List information about all users
    """
    content = {'*': APIMember}

    def get(self):
        return USERS_EXAMPLE


class APIRoot(APIBase):
    """ REST - Access the API as a view
    To test it user:
    curl --user demo:demo  http://localhost:8080/api/members
    curl --user demo:demo  http://localhost:8080/api/members/id1
    """

    # register a REST-method called "members"
    content = {'members': APIMembers}

    def check_authentication(self):
        """ Authenticate requests with basic-authentication user:demo, pwd:demo """
        cred = self.request._authUserPW()
        if cred and len(cred) == 2:
            user, pwd = self.request._authUserPW()
            if user == 'demo' and pwd == 'demo':
                return True

    def get_list_admins(self):
        """ /api/list_admins - list all admin members """
        return [user for user in USERS_EXAMPLE if user['admin'] == True]

    def __init__(self, context, request):
        self.name = "api"
        self.context = context
        self.request = request
